src/operator/src/statement.rs

```diff
@@ -405,6 +450,89 @@ impl StatementExecutor {
                 table_name: table_ref.to_string(),
             })
     }
+
+    async fn plan_exec_inner(
+        &self,
+        stmt: QueryStatement,
+        query_ctx: QueryContextRef,
+    ) -> Result<Output> {
+        let plan = self.plan(&stmt, query_ctx.clone()).await?;
+        self.exec_plan(plan, query_ctx).await
+    }
+}
+
+fn attach_timeout_stream(output: Output, timeout: Duration) -> Output {
+    match output.data {
+        OutputData::AffectedRows(_) | OutputData::RecordBatches(_) => output,
+        OutputData::Stream(stream) => {
+            let stream = TimeoutStream::new(stream, timeout);
+            Output::new(OutputData::Stream(Box::pin(stream)), output.meta)
+        }
+    }
+}
+
+/// A stream to with timeouts.
+struct TimeoutStream {
+    stream: SendableRecordBatchStream,
+    timeout_duration: Duration,
+    start_time: tokio::time::Instant,
+}
+
+impl TimeoutStream {
+    fn new(stream: SendableRecordBatchStream, timeout_duration: Duration) -> Self {
+        TimeoutStream {
+            stream,
+            timeout_duration,
+            start_time: tokio::time::Instant::now(),
+        }
+    }
+}
+
+impl RecordBatchStream for TimeoutStream {
+    fn name(&self) -> &str {
+        self.stream.name()
+    }
+
+    fn schema(&self) -> SchemaRef {
+        self.stream.schema()
+    }
+
+    fn output_ordering(&self) -> Option<&[OrderOption]> {
+        self.stream.output_ordering()
+    }
+
+    fn metrics(&self) -> Option<RecordBatchMetrics> {
+        self.stream.metrics()
+    }
+}
+
+impl Stream for TimeoutStream {
+    type Item = common_recordbatch::error::Result<RecordBatch>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = self.get_mut();
+
+        if this.start_time.elapsed() >= this.timeout_duration {
+            return Poll::Ready(Some(StreamTimeoutSnafu.fail()));
+        }
+
+        Pin::new(&mut this.stream).poll_next(cx)
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.stream.size_hint()
+    }
+}
```

Comment:
I just realized that this stream implementation has a small issue: If the inner stream is pending and waiting for the result, the executor won't wake up the stream.

We should implement this using `stream!` and `select!`. For example:
```rust
stream! {
    let start = Instant::now();
    while let Some(item) = timeout(remaining_timeout, stream.try_next()).await? {
        yield item;
        // TODO: Use checked sub
        remaining_timeout -= Instant::now() - start;
    }
}
```

Reply:
yes i was [concerned](https://github.com/GreptimeTeam/greptimedb/pull/4792#discussion_r1825451682) about this but not sure how to implement. Thanks for the example.

Reply:
Hi @evenyag, I tried to implement the suggestion but failed. I struggled with conversion of interfaces - In the end output accepts a `Pin<Box<dyn RecordBatchStream + Send>>;` . RecordBatchStream inherit from Stream trait and `stream!` return an impl of Stream trait - I am not sure how to convert the input and output generic type between  RecordBatchStream and Stream. Can you give me some guidence?


some of my code is here
```
fn attach_timeout(output: Output, timeout: Duration) -> Result<Output> {
    match output.data {
        OutputData::AffectedRows(_) | OutputData::RecordBatches(_) => Ok(output),
        OutputData::Stream(stream) => {
            let stream = attach_timeout_stream(stream, timeout)?;
            Ok(Output::new(OutputData::Stream(Box::pin(stream)), output.meta))
        }
    }
}

fn attach_timeout_stream<S>(
    mut stream: S,
    mut timeout: Duration,
) -> std::result::Result<impl Stream, common_recordbatch::error::Error>
where
    S: Stream<Item = Result<RecordBatch>> + Unpin
{
    Ok(stream! {
        let start = tokio::time::Instant::now();
        while let Some(item) = tokio::time::timeout(timeout, stream.next()).await.unwrap() {
            yield 
            timeout = timeout.checked_sub(tokio::time::Instant::now() - start).unwrap_or(Duration::ZERO);
        }
    })
}
```

Reply:
You can try this adapter.
https://github.com/GreptimeTeam/greptimedb/blob/d616bd92ef0ca8809ee0387d792714ccbcefb899/src/common/recordbatch/src/lib.rs#L229-L237

Reply:
Hi I am able to get the rust compiler to work with the new impl with stream! macro - i added Send + Unpin train bound to RecordBatchStreamWrapper just fyi.

Reply:
Why do we need `Send` bound in `RecordBatchStreamWrapper`? Can we remove it?

Reply:
removed